Added minimal feature to read/write from/to github repo

shashigharti opened the pull request:
Looking for feedback. I have added minimal tests only and will continue to add more tests.

- fixes #1185

---

Please make sure that all the checks pass. Please add here any additional information regarding this pull request. It's highly recommended that you link this PR to an issue (please create one if it doesn't exist for this PR)

roll committed:
Rebased package validateion on metadata_profile

roll committed:
Rebased schema validation on metadata_profile

roll committed:
Rebased field validation on metadata_profile

roll committed:
Rebased fields validation on metadata_profile

roll committed:
Improved required in metadata_profiles

roll committed:
Fixed additional args in actions

roll committed:
Added minimal resource example

roll committed:
Improved validation error

roll committed:
Recovered strict mode validation

roll committed:
Improved InquiryTask error message

shashigharti committed:
Added required value check (#1178)

Added new feature required-value check
Added tests

roll committed:
Updated to metadata_profile_patch

roll committed:
Enable metadata validation on resource.open

roll committed:
Improved resource.write

roll committed:
Support legacy pipelines

roll committed:
Added support for legacy inquiry

roll committed:
Added frictionless$@4 deprecation warnings

roll committed:
Added name/type patterns to profiles

roll committed:
Required unique field names

roll committed:
Rebased on has_dialect/schema

roll committed:
Improved internal resource logic

roll committed:
Fixed tests

roll committed:
Enabled strict mode for formats.csv

roll committed:
Removed file-level type ignores

roll committed:
Include tests for pyright (the checker is broken currently)

roll committed:
Migrated json parser from create_parser

roll committed:
Migrated jsonl parser from create_parser

roll committed:
Rebased from fullpath to normpath

roll committed:
Index package.analyze/extract by resource.name

roll committed:
Updated dev deps

roll committed:
Implemented resource.normdata

roll committed:
Bootstrapped Catalog class

roll committed:
Bootstrapped Manager API

roll committed:
Bootstrapped portals.ckan

roll committed:
First version of ckanManager.read_package

roll committed:
Support opening package from a manager

roll committed:
Fixed version on catalog from ckan

roll committed:
Support datapackage.json/yaml on ckan

roll committed:
Replaced old ckan implementation

roll committed:
First version of Github manager

roll committed:
Bootstrapped portals.zenodo

roll committed:
Added Metadata.metadata_type

roll committed:
First zenodo implementation

roll committed:
Fixed linting

shashigharti committed:
Added reesource-name argument to extract CLI command (#1194)

Added resource-name argument to CLI command
Added tests

roll committed:
Fixed timezone issue (#1197)

roll committed:
Fixed package.from_ckan

roll committed:
Merge branch 'main' into v5

roll committed:
Fixed field_filter step

roll committed:
Fixed tests

roll committed:
Fixed tests

roll committed:
Skip test with pandas warning

roll committed:
Skip failing tests on windows

shashigharti committed:
Fixed missing required info in resource summary (#1193)

* Fixed missing required info in resource summary

Fixed missing info of required property in resource summary
Added tests

* Revised code

Removed extra code to check if required field exists
Added error message to notify invalid Field property required
Added new tests

* Changed code to align with v5 changes

roll committed:
Fixed bigquery timezone issue

roll committed:
Removed debug print

roll committed:
Fixed tests

roll committed:
Rebased on platform class for extras

roll committed:
Return excel to extras

roll committed:
Return json into extras

roll committed:
Removed is_platform from tests

roll committed:
Fixed platform.type

roll committed:
Moved api to extras

roll committed:
Added platform.jinja2

roll committed:
Added platform.petl

roll committed:
Removed package imports from init

roll committed:
Added platform.chardet

roll committed:
Added platform.jsonschema/yaml

roll committed:
Added platform.validators/rfc3986

roll committed:
Added platform.dateutil_parser/isodate

roll committed:
Removed backport for compression exceptions

roll committed:
Rebased fields.geojson on platform

roll committed:
Added platform.zipfile/gzip

roll committed:
Dropped jsonmerge dependency

roll committed:
Added platform.requests

roll committed:
Removed helpers.is_platform/is_python

roll committed:
Added platform.marko/html_parser

roll committed:
Drop yml extension support

roll committed:
Added inquiryTask.strict prop

roll committed:
Improved system error messages

roll committed:
Stick to SHA256 hashing

roll committed:
Fixed report.from_descriptor

roll committed:
Make resource.stats an object

roll committed:
Make report/reportTask.stats an object

roll committed:
Fixed tests

roll committed:
Fixed tests

roll committed:
Updated stats.bytes on reading

roll committed:
Support MD5 hash as well

roll committed:
Updated errors

roll committed:
Fixed tests

roll committed:
Renamed stats.time -> seconds

roll committed:
Removed resource.__iter__

roll committed:
Fixed v0/url rule

roll committed:
Create dialect/stats in the getters

shashigharti committed:
Implemented program.convert (#1195)

Included additional metadata classes
Added markdown template for metadata classes whose markdown template were missing
Added tests

roll committed:
Update convert feature to the latest changes

roll committed:
Fixed program.convert not found logic

roll committed:
Improved metadata logic

roll committed:
Support Parquet data format (#1186)

* Implemented parquet reading

* Implemented parquet writing

* Fixed linting

* Updated comment

* Updated supported_types; rebase on platform

roll committed:
Fixed ParquetControl

roll committed:
Added a issue-1205 test

shashigharti committed:
Added doc blocks to the cell steps classes (#1204)

Added docblocks to steps classes

roll committed:
Fixed linting

roll committed:
Fixed tests

roll committed:
Updated tests

roll committed:
Added dereference tests

roll committed:
Added safety tests

roll committed:
Fixed behaviour on emty resources

roll committed:
Fixed csv quote issue

roll committed:
Fixed more tests

roll committed:
Fixed more package tests

roll committed:
Removed debug print

roll committed:
Fixed type for html format

roll committed:
Updated xfails

roll committed:
Review actions' xfails

roll committed:
Updated xfails

shashigharti committed:
1205/recover failing tests (#1209)

* Fixed failing tests

Fixed tests for package.dot, markdown table and convert feature

* Code cleanup

* Fixed tests in schema->test_convert

* Fixed tests in inquiry->test_convert

* Fixed tests in analyze->test_general.py

* Fixed tests in program->test_convert and program->test_summary

* Fixed tests in program->test_validate

* Fixed tests in program->test_validate and report->test_convert

* Fixed failing tests in program->test_validate

* Restored resource file

roll committed:
Fixed linting

roll committed:
Fixed tests

roll committed:
Fixed custom-field

roll committed:
Enabled xpassed tests

roll committed:
Fixed extract-dialect

roll committed:
Recovered spss tests

roll committed:
Fixed sync-schema

roll committed:
Recovered inquiry ci tests

roll committed:
Fixed gsheets on ci

roll committed:
Fixed bigquery tests

roll committed:
Fixed bigquery on ci

roll committed:
Fixed bigquery on ci

roll committed:
Fixed bigquery on ci

roll committed:
Fixed bigquery on ci

roll committed:
Fixed bigquery on ci

roll committed:
Skip ckan tests until 475 is done

roll committed:
Fixed report tests

roll committed:
Fixed linting

roll committed:
Fixed linting

roll committed:
Updated skip reason

roll committed:
Fixed tests

roll committed:
Skip bigquery tests

roll committed:
Handled TODOs in resource tests

roll committed:
Refactored metadata.metadata_validate

roll committed:
Implemented resource.profiles

roll committed:
Implemented package.profiles

roll committed:
Fixed package-zip

roll committed:
Fixed json-data

roll committed:
Replaced docs TODO by NOTE

roll committed:
Fixed resource's safety

roll committed:
Fixed package's safety

roll committed:
Updated dereference test

roll committed:
Recovered steps.field_update

roll committed:
Recovered field-update related tests

roll committed:
Removed outdated NOTEs

roll committed:
Recovered steps.field_add

roll committed:
Recovered steps.field_merge

roll committed:
Recovered steps.field_pack

roll committed:
Recovered steps.field_split

roll committed:
Recovered steps.resource_update

roll committed:
Renamed metadata to descriptor in steps

roll committed:
Recovered steps.row_sort

roll committed:
Recovered steps.row_subset

roll committed:
Recovered steps.table_write

roll committed:
Skip steps.table_pivot

roll committed:
Recovered steps.table_attach

roll committed:
Recovered steps.table_diff

roll committed:
Recovered steps.table_intersect

roll committed:
Skip steps.table_merge tests

roll committed:
Recovered steps.table_join

roll committed:
Fixed dereference for metadata

roll committed:
Recovered resource's dereference

roll committed:
Back to metadata.to_descriptor_source

roll committed:
Recovered package's dereference

roll committed:
Added new metadata validate API

roll committed:
Use explicit terminology for legacy

roll committed:
Fixed mediatype

roll committed:
Make json non-tabular by default

roll committed:
Support non-tabular resource.data

roll committed:
Minor resource code update

roll committed:
Added fk issue

roll committed:
Handle schema.fields not array

roll committed:
Handle crictial descriptor errors in structure

roll committed:
Fixed field validation

roll committed:
Catch unsupported field types

roll committed:
Removed incorrect tests

roll committed:
Fixed data is a string error catching

roll committed:
Bootstrapped metadata's API@2

roll committed:
Started rebasing Schema on new metadata API

roll committed:
Recovered schema tests

roll committed:
Recovered fields tests

roll committed:
Started resource migration

roll committed:
Recovered some resource tests

roll committed:
Recovered schema/dialect tests

roll committed:
Recovered more resource tests

roll committed:
Started recovering resource.validate

roll committed:
Recovered resource tests

roll committed:
Rebased on metadata_specify

roll committed:
Moved metadata internals to helpers

roll committed:
Migrated package

roll committed:
Fixed linting

roll committed:
Added square brackets test

shashigharti committed:
Added minimal feature to read/write from/to github repo

Added read/write and search
Added few tests

roll commented:
Hi @shashigharti, in general I think you made great progress! I'll be commenting in the PR

roll commented on tests/portals/github/test_manager.py:
```diff
@@ -0,0 +1,154 @@
+# Write
+
+
+@pytest.mark.vcr
```
There should be a VCR option to exclude sensitive information from the cassete. See this commit - https://github.com/frictionlessdata/frictionless-py/blob/7db069431b2f5f60c518e5c05dfedafa53504f28/data/cassettes/test_github_parser_write_package_file.yaml - our Github Token is a part of the cassete. Nothing to worry about as Github automatically revoked this token when it hit the public repo but we need a way to exclude it from the cassete

roll commented on tests/portals/github/test_manager.py:
```diff
@@ -0,0 +1,154 @@
+
+
+@pytest.mark.vcr
+def test_github_parser_read(options_without_dp):
```
test_github_manager_read

roll commented on tests/portals/github/test_manager.py:
```diff
@@ -0,0 +1,154 @@
+    user = options_without_dp.pop("user")
+    repo = options_without_dp.pop("repo")
+    control = portals.GithubControl()
+    manager = portals.GithubManager(control)
```
In the tests we need to use the Package class instead of the Manager class as per requirements. Manager is an internal class; the framework uses under the hood so the user would write (both should work):

```
package = Package(control=portals.GithubControl(user=user, repo=repo))
package = Package('https://github.com/user/repo")
```

roll commented on frictionless/portals/github/control.py:
```diff
@@ -15,14 +15,38 @@ class GithubControl(Control):
+    page: Optional[int] = None
+    """NOTE: add docs"""
+
+    apikey: Optional[str] = None
```
If you use ENV value here as a default you don't need to handle it anywhere else just use `control.apikey`

```
apikey: Optional[str] = os.environ.get("GITHUB_ACCESS_TOKEN")
```

roll replied:
The same for email

roll commented on frictionless/portals/github/manager.py:
```diff
@@ -1,45 +1,156 @@
+        print(search)
+        assert search
+        if self.control.page:
+            search = f"{search}&page={self.control.page}"
```
Maybe using Python's builtin means for building query string. It's not really important but you can try searching for urlib.parse for query strings (can't remember the function)

shashigharti replied:
I have removed the query string generation part.

roll commented on frictionless/portals/github/manager.py:
```diff
@@ -1,45 +1,156 @@
+    return paths
+
+
+def _get_package(paths, repository, base_path, formats):
```
It's OK not to use the leading underscore here as it's an internal function anyway. We don't use the leading underscore for cases like this across the codebase

roll commented on frictionless/portals/github/manager.py:
```diff
@@ -1,45 +1,156 @@
+        return response(repository, file_created)
+
+
+def _get_contents(contents, repository):
```
I would make this function `get_resources() -> List[Resource]` as on the next step we also need to add some metadata like `resource.name`. Probably the best way for doing so will be:

```
resource = Resource(path=path)
resource.infer(sample=False) # it means infer anything you can without any file reading
```

roll replied:
But if Github API provides metdata like scheme/format/mediatype etc we need to set it directly

```
resource = Resource(path=path, format=format, ...)
```

shashigharti replied:
I used resource.infer as I couldn't find any function that would return these file metadata.

get_content is returning all the paths that also includes datapackage.json or datapackage.yaml so was not sure if it could be included as resource get_resources() -> List[Resource]. So I changed the name of the function but return type set as list[str]

roll commented on tests/portals/github/test_manager.py:
```diff
@@ -0,0 +1,154 @@
+    control = portals.GithubControl(search="test user:fdtester")
+    manager = portals.GithubManager(control=control)
+    apikey = os.environ.get("GITHUB_ACCESS_TOKEN", None)
+    packages = manager.read_catalog(apikey=apikey)
```
Note that `manager.read_catalog() -> Catalog` needs to return the Catalog class. So in tests we write:
```
catalog = Catalog(control=control)
for package in catalog.packages:
    ...
```

roll commented on frictionless/portals/github/control.py:
```diff
@@ -15,14 +15,38 @@ class GithubControl(Control):
     repo: Optional[str] = None
     """NOTE: add docs"""
 
+    formats: Optional[list] = [".csv"]
```
In our terminology format doesn't contain dots
```
formats: Optional[List[str]] = ['csv']
```

roll commented on data/fixtures/portals/github/datapackage.json:
```diff
@@ -0,0 +1,34 @@
+{
```
I think we don't need to put the test packates under `fixtures/portals/github` we can just have `data/package-name` and use them across the integrations (ckan/github/zenodo/etc)

roll replied:
I would also have at least 2 resources there

shashigharti replied:
I removed those files and used the existing files datapackage.json.

It has two resources in it.

roll committed:
Recovered resource tests

roll committed:
Recovered package/format tests partially

roll committed:
Recovered actions.describe

roll committed:
Recovered steps

roll committed:
Recovered Inquiry tests

roll committed:
Fixed metadata tests

roll committed:
Recovered actions

roll committed:
Simplified actions code

roll committed:
Improved error catching in validate

roll committed:
Recovered error-catching tests partially

roll committed:
Recovered format tests

roll committed:
Added more xfails

roll committed:
Recovered program tests

roll committed:
Recovered package.validate tests

roll committed:
Xfailed tests

roll committed:
Fixed report tests

roll committed:
Migrated Error to new metadata API

roll committed:
Removed reportTask.scope

roll committed:
Fixed report's stats

roll committed:
Removed system.create_field

roll committed:
Removed system.create_error

roll committed:
Removed system.create_step

roll committed:
Removed system.create_check

roll committed:
Removed system.create_control

roll committed:
Renamed detect_field_candidates

roll committed:
Recovered profiles steps

roll committed:
Rebased on exception.to_errors()

roll committed:
Recovered program.transform tests

roll committed:
Imporved transform comments

roll committed:
Fixed strict mode

roll committed:
Fixed zip tests

roll committed:
Migrated on system.trusted

roll committed:
Recovered security tests

roll committed:
Fixed package.to_zip

roll committed:
Rebased on system.use_context

roll committed:
Rebased on system.onerror

roll committed:
Removed package.detector

roll committed:
Improved metadata.metadata_import

roll committed:
Fixed dereference tests

roll committed:
Reworked basepath logic

roll committed:
Minor improvements

roll committed:
Fixed print(report)

roll committed:
Added resource.write(control)

roll committed:
Merge branch 'main' into v5

roll committed:
Fixed deps

roll committed:
Fixed tests

shashigharti committed:
Added catalog, new tests  and revised code

1. Added catalog search and tests
2. New tests for to_github
3. New tests for from_github
4. Added from_github and to_github and publish
5. Added fixture to hide/filter token in header

shashigharti committed:
Fixed merge conflict and failing tests

roll commented on frictionless/package/package.py:
```diff
@@ -441,6 +441,62 @@ def to_ckan(self, target, *, control=None):
+        """
+        manager = system.create_manager(source, control=control)
+        if not manager:
+            raise FrictionlessException(f"Not supported Github source: {source}")
```
Maybe also adding control to the error note as source might be None

roll commented on frictionless/package/package.py:
```diff
@@ -441,6 +441,62 @@ def to_ckan(self, target, *, control=None):
         return storage
 
+    @staticmethod
+    def from_github(
```
What if this function will just use constructor?

I think it will be the same:
```
return Package(source, control=control)
```

WDYT?

aivuk replied:
I agree with that. Better to keep all the Catalog specific logic inside the portals/github

roll commented on frictionless/package/package.py:
```diff
@@ -441,6 +441,62 @@ def to_ckan(self, target, *, control=None):
+        package = manager.read_package()
+        return package
+
+    def to_github(
```
What if this function will just call `publish`?

I think it will be the same:
```
return self.publish(source, control=control)
```

roll reviewed (commented):
Thanks @shashigharti!

It already looks realy solid I'll continue reviewing when you add resource writing to Github

---

> I added publish function but both to_github and publish has same functionality 'write_package' should I add new function for publish in manager.

I commented in the review regarding this

> Write function only writes datapackage.json file. Shall I add function to write resources?

Yes. We need to publish resource as well. The user story here is that the use will be able upload the whole data package to a github repo 

> I haven't yet added tutorial, I will tutorial once the functionality is completed and approved.

I'm working on the docs improvements now so you will be able to write a tutorials easier than we do now. I'll push changes around tomorrow and we'll discuss

roll commented:
Hi @aivuk can you please also review/test/participate when you have time?

roll commented on frictionless/portals/github/manager.py:
```diff
@@ -1,45 +1,190 @@
+        packages = []
+        query: Dict[str, Any] = {}
+
+        if self.control.repo or self.control.user or self.control.search:
```
I would suggest raising an exception here like:

```
if not self.control.repo and not ...:
    raise
```

It will allow to write the following code without indentation (just slightly easier to read and maintain)

roll commented on frictionless/portals/github/manager.py:
```diff
@@ -1,45 +1,190 @@
+                base_path = f"https://raw.githubusercontent.com/{location}/{repository.default_branch}"
+                contents = repository.get_contents("")
+                resource_path = get_resources(contents, repository)
+                packages = get_packages(resource_path, base_path)
```
We can use some data from `control` as a catalog name e.g. a github user name

roll commented on frictionless/portals/github/manager.py:
```diff
@@ -1,45 +1,190 @@
+                raise FrictionlessException(note)
+
+        response = namedtuple("response", ["repository", "file_created"])
+        return response(repository, file_created)
```
I think here and in `Packge.to_github` we need to return a full URL to the published data package like: https://raw.githubusercontent.com/fdtester/test-repo-with-datapackage-yaml/master/datapackage.yaml

roll replied:
But I'm not yet sure. WDYT?

shashigharti replied:
I think returning the package file makes sense. Or may be returning the Repository only and we can use this method of `Repository` to test the package file. 
`repository.get_contents(package_filename)`

roll commented on tests/portals/github/test_manager.py:
```diff
@@ -0,0 +1,404 @@
+    url = options_without_dp.pop("url")
+    output = options_without_dp.pop("output")
+    package = Package(url)
+    assert json.loads(package.to_json()) == output
```
`json.loads(package.to_json())` -> `package.to_descriptor` :wink:

roll commented on frictionless/portals/github/control.py:
```diff
@@ -12,17 +13,57 @@ class GithubControl(Control):
+    email: Optional[str] = os.environ.get("GITHUB_EMAIL", None)
+    """NOTE: add docs"""
+
+    formats: Optional[List[str]] = ["csv"]
```
Let's have ['csv', 'xlsx', 'xls'] by default and also we might have it as DEFAULT_FORMATS const

roll commented on frictionless/portals/github/control.py:
```diff
@@ -12,17 +13,57 @@ class GithubControl(Control):
+    page: Optional[int] = None
+    """NOTE: add docs"""
+
+    per_page: Optional[int] = 30
```
Maybe DEFAULT_PER_PAGE const to the top of the file

roll commented on tests/portals/github/test_manager.py:
```diff
@@ -0,0 +1,404 @@
+                "type": "table",
+                "path": "https://raw.githubusercontent.com/fdtester/test-repo-without-datapackage/master/data/student.xlsx",
+                "scheme": "https",
+                "format": "xlsx",
```
It will be great if we can have `encoding` here also. Is it possible to get it from Github API metadata?

roll commented on tests/portals/github/test_manager.py:
```diff
@@ -0,0 +1,404 @@
+
+
+@pytest.mark.vcr
+def test_github_manager_catalog_from_repo_with_multiple_packages(
```
@shashigharti 
Let's keep it simple: one repo == one data package so we don't need to create a catalog from one repo

roll commented:
@shashigharti 
Another test suggestion:
```
catalog = Catalog("https://github.com/fdtester")
print(catalog)
```
Currently, it fails for me I guess because we need to de-duplicate package names. I think we might run into the same problem for resource if we have a repo like this:
```
data/table.csv
table.csv
```

BTW can we also have a test repo like this (with multiple files without a data package):

```
legacy/table.xls
table.csv
table.xlsx
```

shashigharti committed:
Revised code and added new tests

Fixed resource writing to github
Added type hint
Removed get_packages function (assumption: one repo will only have one package file)
Fixed failing tests
Added new tests

shashigharti commented:
@roll @aivuk Thank you very much for the feedback!. I have made the changes as per your feedback and it is ready for the review.

I also made some changes to the repo because of duplicate resources and changed previous failing test.

roll commented:
Thanks @shashigharti! I'll ve rewewing it today or tomorrow

shashigharti committed:
Fixed bug for base path, fixed test

shashigharti commented:
Closing this PR and creating a new one.
